grayscale example simd

mehmetyusufoglu opened the pull request:
Grayscale value calculation from RGB values. R,G and B are 1D arrays, result is grayScaleValues array. Simd feature is used.

mehmetyusufoglu committed:
grayscale example

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,211 @@
+    onHost::Queue queue = devAcc.makeQueue();
+
+    // Define the work division
+    IdxVec const extent(1024 * 1024);
```
please read the problem size from the command line. I will change it for the evctorAdd too.

mehmetyusufoglu replied:
done, thanks.

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+
+// Define global constants for grayscale conversion
+constexpr uint32_t scalarR = 11u;
+constexpr uint32_t scalarG = 5u;
```
not importent for performcne but the waitings are IMO wrong

should be `16u`

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+// Define global constants for grayscale conversion
+constexpr uint32_t scalarR = 11u;
+constexpr uint32_t scalarG = 5u;
+constexpr uint32_t scalarB = 16u;
```
should be `5`

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+}
+
+// SIMD Kernel class template
+class GrayscaleCalculationSimdKernel
```
should be `GrayscaleKernel`

mehmetyusufoglu replied:
Done

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+    }
+}
+
+// SIMD Kernel class template
```
string can be removed

mehmetyusufoglu replied:
done

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+    GrayscaleCalculationSimdKernel kernel;
+
+    // Define chunk size and compute blocking specification
+    Vec<size_t, 1u> chunkSize = 256u; // Number of threads per block
```
`chunkSize` -> `frameExtent`

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+    // Define chunk size and compute blocking specification
+    Vec<size_t, 1u> chunkSize = 256u; // Number of threads per block
+    uint32_t elementsPerWorker = alpaka::getNumElemPerThread<Data>(alpaka::onHost::getApi(queue));
+    auto dataBlocking = onHost::FrameSpec{divCeil(extent, chunkSize * elementsPerWorker), chunkSize};
```
could you change it to 

```
auto dataBlocking = onHost::FrameSpec{divExZero(extent, chunkSize * elementsPerWorker), chunkSize};
```

mehmetyusufoglu replied:
There is a problem at divExZero function. No return type.

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+
+// Template function for fuzzy equality
+template<typename T>
+[[maybe_unused]] bool FuzzyEqual(T a, T b)
```
please remove this function, not required for intergral types

mehmetyusufoglu replied:
removed.

psychocoderHPC commented on example/grayScaleSimd/src/grayScaleSimd.cpp:
```diff
@@ -0,0 +1,207 @@
+             scalarB * bufHostB[i]) // Direct indexing for B
+            / scalar32); // Normalize by scalar32
+
+        if(!FuzzyEqual(computedGray, correctResult))
```
we use integral types please remove fuzzy comparisn

psychocoderHPC committed:
renmae exmaple `grayScaleSimd` to `grayScale`

psychocoderHPC committed:
gra scale: refactorings and fixes

- fix implicit used 32bit index for the validation loop
- fix validation, only parts ouf the 32bit output value was checked
- fix not initialized alpaha channel
- fix loading the pixel value multiple times in the kernel
- beautify output
- add useful information to output
- add license to example

psychocoderHPC commented:
@mehmetyusufoglu I reworked the example

psychocoderHPC committed:
refactor gray scale example

- change the code to performe a inplace color transformation
- add the native algorith which is using `std::for_each()`
- add README and reference Mattihas Kretz `std::simd` talk
